File: kotlin-node/karakum/src/jsMain/resources/plugins/convertUndiciTypesReexport.js
import ts from "typescript";

export default function (node) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("http.d.ts")) return null

    if (
        ts.isVariableDeclaration(node)
        && ts.isIdentifier(node.name)
        && (
            node.name.text === "WebSocket"
            || node.name.text === "CloseEvent"
            || node.name.text === "MessageEvent"
        )
    ) {
        return ""
    }
    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertUndiciTypesReexport.kt
package node.karakum.plugins

import node.karakum.util.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isIdentifier
import typescript.isVariableDeclaration

val convertUndiciTypesReexport = createPlugin { node, context, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("http.d.ts"))

        ensure(isVariableDeclaration(node))

        val name = node.name
        ensure(isIdentifier(name))
        ensure(
            name.text == "WebSocket" ||
                    name.text == "CloseEvent" ||
                    name.text == "MessageEvent"
        )

        ""
    }
}
